### PR TITLE
779 feature export student results in html

### LINKED ISSUE #779: Export student results to html
Implement feature for export student results in html format
(/api/v2/exports/studentResults/{studentId}/{extension})

### PR DESCRIPTION
**Pull request by BobTheSuperman**
Added the ability to get the average student score and the average number of visits in html format

### CODE REVIEW

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/HtmlExportService.cs [hunk 1]
```diff
@@ -71,11 +71,18 @@ public Task<Result<FileDto>> GetStudentGroupResults(StudentGroupsResultsDto data
                     "student classbook can't be returned in html format"));
         }
 
-        public Task<Result<FileDto>> GetStudentResults(StudentsResultsDto data)
+        public async Task<Result<FileDto>> GetStudentResults(StudentsResultsDto data)
```
Comment (DmitryTkachenko) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/HtmlExportService.cs]: Please add Async

**Review by DmitryTkachenko — CHANGES_REQUESTED** [review]

**Review by Ssashaa — APPROVED** [review]

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/HtmlExportService.cs (hunk 1 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/HtmlExportService.cs]: Done

**Review by apyura — APPROVED** [review]

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        private void FillFile(StudentsResultsDto data)
+        {
+            if (data == null)
+            {
+                return;
+            }
+
+            var table = new StringBuilder();
+
+            bool averageVisitsExistAndValid = data.AverageStudentVisits != null && data.AverageStudentVisits.Any();
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: please add list initialization in dto and remove null check

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        #region private constants
+
+        private const int CourseColumnNumber = 0;
+        private const int StudentGroupColumnNumber = 1;
+        private const int ThirdColumnNumber = 2;
+        private const int FourthColumn = 3;
+
+        #endregion
+
+        private void FillFile(StudentsResultsDto data)
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs:33]: please move private methods after public

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+            var table = new StringBuilder();
+
+            bool averageVisitsExistAndValid = data.AverageStudentVisits != null && data.AverageStudentVisits.Any();
+
+            bool averageMarksExistAndValid = data.AverageStudentsMarks != null && data.AverageStudentsMarks.Any();
+
+            if (averageMarksExistAndValid && averageVisitsExistAndValid)
+            {
+                IEnumerable<AverageStudentMarkDto> studentMarks = data.AverageStudentsMarks;
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: please use to list()

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            string student = string.Empty;
+
+            if (data.AverageStudentVisits != null && data.AverageStudentVisits.Any())
+            {
+                student = data.AverageStudentVisits.First().Student;
+            }
+            else if (data.AverageStudentsMarks != null && data.AverageStudentsMarks.Any())
+            {
+                student = data.AverageStudentsMarks.First().Student;
+            }
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: please simplify (use ternary operator)

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+            return $"Student: {student} average results";
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: please use move to constatns

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        public override async Task FillFileAsync(StudentsResultsDto data)
+        {
+            await Task.Run(() => FillFile(data));
+        }
+
+        public override string GetFileName()
+        {
+            return "StudentResults_" + DateTime.Now.ToString("yyyy-MM-dd") + ".html";
```
Comment (ArsenMaistryshyn) [inline CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: please move to constatns

**Review by ArsenMaistryshyn — CHANGES_REQUESTED** [review]

**Review by DmitryTkachenko — APPROVED** [review]

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs (hunk 2 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs (hunk 4 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs (hunk 6 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs (hunk 7 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs (hunk 5 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs]: Done!

File: CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs (hunk 3 above)
Comment (BobTheSuperman) [reply CharlieBackend.Business/Services/FileServices/ExportFileServices/Html/StudentResultsHtmlFileExport.cs:33]: Done

**Review by ArsenMaistryshyn — APPROVED** [review]

**Pull request merged**